src/ray/util/internal/tests/stream_redirection_utils_test.cc

```diff
@@ -0,0 +1,137 @@
+// Copyright 2025 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ray/util/internal/stream_redirection_utils.h"
+
+#include <gtest/gtest.h>
+
+#include <chrono>
+#include <iostream>
+#include <string>
+#include <thread>
+#include <vector>
+
+#include "ray/common/test/testing.h"
+#include "ray/util/filesystem.h"
+#include "ray/util/util.h"
+
+namespace ray::internal {
+
+namespace {
+
+// Output logging files to cleanup at process termination.
+std::vector<std::string> log_files;
+void CleanupOutputLogFiles() {
+  for (const auto &cur_log : log_files) {
+    EXPECT_TRUE(std::filesystem::remove(cur_log));
+  }
+}
+
+}  // namespace
+
+class LoggingUtilTest : public ::testing::Test {
+ public:
+  void SetUp() override {
+    // Cleanup generated log files at test completion; because loggers are closed at
+    // process termination via exit hook, and hooked functions are executed at the reverse
+    // order of their registration, so register cleanup hook before logger close hook.
+    ASSERT_EQ(std::atexit(CleanupOutputLogFiles), 0);
+  }
+};
+
+TEST(LoggingUtilTest, WriteContentWithNewliner) {
+  constexpr std::string_view kLogLine1 = "hello\n";
+  constexpr std::string_view kLogLine2 = "world\n";
+
+  const std::string test_file_path = absl::StrFormat("%s.err", GenerateUUIDV4());
+  const std::string log_file_path1 = test_file_path;
+  const std::string log_file_path2 = absl::StrFormat("%s.1", test_file_path);
+  log_files.emplace_back(log_file_path1);
+  log_files.emplace_back(log_file_path2);
+
+  // Works via `dup`, so have to execute before we redirect via `dup2` and close stderr.
+  testing::internal::CaptureStderr();
+
+  // Redirect stderr for testing, so we could have stdout for debugging.
+  StreamRedirectionOption opts;
+  opts.file_path = test_file_path;
+  opts.tee_to_stderr = true;
+  opts.rotation_max_size = 5;
+  opts.rotation_max_file_count = 2;
+  auto redirection_handle = RedirectStreamImpl(GetStderrHandle(), opts);
+
+  std::cerr << kLogLine1 << std::flush;
+  std::cerr << kLogLine2 << std::flush;
+
+  // TODO(hjiang): Current implementation is flaky intrinsically, sleep for a while to
+  // make sure pipe content has been read over to spdlog.
```

Comment:
Can you help me understand the path to making it deterministic? Is it "impossible" or just requires some more work?

we should avoid sleeps in unit tests as much as possible for two reasons:
1. it makes them intrinsically flaky (as you noted)
2. it makes the unit tests slow. In an ideal world, for developer productivity we would be able to run the full suite of C++ unit tests in the codebase very quickly. Having individual tests with multi-second sleeps gets in the way of that.

Reply:
At a minimum, you should be able to mitigate both of these issues by restructuring the test to repeatedly check for the expected behavior on a short interval (or in a busy loop) and only fail if the expected behavior is not observed after a longer timeout.

This is a common pattern in other Ray tests.

Reply:
> Can you help me understand the path to making it deterministic? Is it "impossible" or just requires some more work?

A lot more work.

In terms of effort to make it deterministic, you need to
- Write a special indicator to a uni-directional pipe in thread thread (main thread, thread-1);
- Listen thread (thread-2) checks whether received string contains the indicator;
- Listen thread asks dump thread (thread-3) to do a flush;
- The tricky thing is we use uni-directional pipe (since in production there's no meaning for the reserve direction) for main thread and background threads, there's no directly way to echo back;
- You could use filesystem for IPC (i.e. write a file) but it takes busy polling which is also not good

Reply:
> it makes the unit tests slow

Yes. Current testing for ray, a C++ test pipeline usually takes several hours, within which C++ unit testing itself only takes minutes.

Reply:
> At a minimum, you should be able to mitigate both of these issues by restructuring the test to repeatedly check for the expected behavior on a short interval (or in a busy loop) and only fail if the expected behavior is not observed after a longer timeout.

I doubt it.

We use gtest to capture stdout and stderr, 
https://github.com/google/googletest/blob/main/googletest/src/gtest-port.cc#L1201-L1209

stream object gets deleted after one invocation, cannot be repeated called.

Reply:
> Yes. Current testing for ray, a C++ test pipeline usually takes several hours, within which C++ unit testing itself only takes minutes.

CI is one consideration, but one of the main motivations for making unit tests fast is for rapid iteration locally. If we can run a large swath of unit tests quickly on our laptops, it increases developer productivity. I have worked in codebases where this is done well and it's very empowering.

And yes, while the current state of the C++ tests and CI is poor, we should be striving to make it better and set a new standard with each added PR

Reply:
I totally agree, I illustrate the difficulty to make it unit testable above.
Do you want me to add the reasoning to the comment?